### ISSUE TITLE
Filter problems in the Review field of the course notes

### ISSUE THREAD

**Issue by florian-rabe** [slatex/sTeX-React#463]
Currently, all problems are shown here.

It should be possible to filter by, e.g., quiz/homework problems from this/previous semester.
That would point the students to the most relevant, up-to-date, and best-annotated problems.

**Comment by abchugh** [comment]
The frontend would happily do that if we can get such data from the MMT/iMMT backend. As far as I know we can only filter based on the a few things like 
 
 @Jazzpirate can confirm what's possible and if there are any potential easy solutions for getting this metadata (Was it used in a homework or quiz? If so, in which semester?)  about a problem. 

Another approach would be for instructors (like you) to be given a complete list of problems and you would be able to accept /reject the problems that are good enough for the "Review field". The instructor could:
1. Add some sort of annotation/marker in the problem sources (sTeX) OR
2. The list of appropriate problems could be stored in MMT/LMS/BoFE.
Although, this could be a step against the spirit of ALeA. (Instructors should need to do minimal work after authoring, the system should be able to determine which LOs should be shown on it own based on provided data)

Another consideration is that we eventually want LMS to decide which problems are to be shown based on the learner model, so this decision of what problems are to be shown should be made there. Assigning to @lambdaTotoro for further thoughts.

**Comment by lambdaTotoro** [comment]
I am assuming it would be possible to access all instances of a given course and collect what problems have been used via the JSON files that were submitted for homeworks and quizzes.
Once that is available to LMP in some way (e.g. as a SPARQL verb) I can incorporate the information in the (already available) [learning object recommendation](https://gitos.rrze.fau.de/voll-ki/fau/system/alea-lms/-/blob/main/docs/API/LearningObjects.md?ref_type=heads) API or a future iteration thereof.

But the sort of filter-by-metadata that Florian asked for is clearly a content issue and fundamentally unrelated to learner modelling. Unassigning.

**Comment by Jazzpirate** [comment]
> The frontend would happily do that if we can get such data from the MMT/iMMT backend. As far as I know we can only filter based on the a few things like
> 
> @Jazzpirate can confirm what's possible and if there are any potential easy solutions for getting this metadata (Was it used in a homework or quiz? If so, in which semester?) about a problem.

"course", "instance", "homeworks" and "quizzes" are not (i)MMT-concepts. The information "this problem was used in this course in that year as a homework in this week" is very much a thing that (i)MMT knows nothing about, and I suspect it shouldn't either; so a SPARQL query to (i)MMT isn't the answer. Moreover, currently, via the JSON uploads, homeworks and quizzes circumvent the (i)MMT backend entirely - i.e. it may not even know the problems themselves; let alone when they were used where.

The only system that knows that metadata is the BOF ;)

**Comment by kohlhase** [comment]
Actually, I think that with enough metadata in the sTeX sources (e.g. `this document is an AI-1 exam that was given in WS2324` sTex-encoded and exported to the RDF would solve this problem.

**Comment by Jazzpirate** [comment]
Sure, ultimately, any information can be hardcoded as additional sHTML attributes using additional sTeX macros. I'm questioning whether that is smart from a separation-of-concerns-POV though.
I consider courses, course instances, specific homework sets / tuesday quizzes in specific course semester etc. ALeA functionality, as opposed to "generic semantically annotated documents"(=iMMT/sTeX) functionality. Some ALeA functionality may require dedicated iMMT/sTeX functionalities, but by and large, I think it's a good idea to keep them conceptually separate as much as possible.
And, importantly, ALeA already has the information, without the need of manually annotating it somewhere ;)

**Comment by kohlhase** [comment]
I agree that this information should be hosted and managed on ALeA. And that is not so much. ALeA already knows what the quizzes are, and what problem (files) are referenced. It currently does not know this information about exams, but should. That would be enough to mark problems as quiz, homework, and exam problems (for the current course) and thus do what Florian has requested. I will make an infrastructural issue for that.

**Comment by achugh-wald** [comment]
Regarding keeping the data related to quiz, homework, exam in BoF: sounds good to me.

BUT, the information regarding quizzes and Homework is irrelevant. For quizzes and homeworks, we upload JSONs and not pointers to MMT uris. So we can still not do anything with this info on the "review section" - as requested by Florian.

For exams, the stored information will be useful. BUT, the "review section" problems are (soon?) going to be decided based on learner models. It feels wrong to do a post-filter based on the learning objects returned by LMP. Instead, it would architecturally make more sense if LMP queries BoF for this info and returns the appropriate problems directly.

Thoughts, @lambdaTotoro ?

**Comment by kohlhase** [comment]
> For quizzes and homeworks, we upload JSONs and not pointers to MMT uris. So we can still not do anything with this info on the "review section" - as requested by Florian.

you are right, this needs to change, we also need the URLs. Well spotted. This might be quick for @Jazzpirate to retrofit or to remember in the new backend.

**Comment by Jazzpirate** [comment]
The JSON-upload-workflow may or may not be needed with the new backend anymore... Ideally, it wouldn't be, but I can imagine that pushing, rebuilding and possibly iterating because of only-tangentially-related build errors is too much overhead when a new quiz is due in 15 minutes - so let's assume we keep the json workflow :)
Then adding URIs for the problems to the JSON is easy *but* the backend will not now about those problems unless and until they are pushed to the backend and are not being moved around in the meantime ;)
URIs are still only persistent relative to archive+relative path and all that.

Having the BoF query the LMP, only to have the LMP query the BoF again for the metadata, smells very strongly to me. My suspicion (NEEDS DESIGN[TM]) is that we *either* want to filter by course *or* by learner model. If we want "here are meaningful course-relevant practice problems for this section based on homeworks, previous exams etc.", then we don't want user model filtering, and if we want the latter, we don't want the former. Doing both feels conflicting not just in terms of the not-yet-existent filtering algorithm, but also in terms of user-expectation what kind of exercises they should see...?

**Comment by abchugh** [comment]
Minor correction in @Jazzpirate's comment: the react frontend (and not BoF) will query LMP which in turn will query BoF. 

But you still may object since the frontend and BoF are somewhat coupled together.

**Comment by Jazzpirate** [comment]
Fair point, but yeah, my objection is less about it being a roundtrip rather than having two entirely distinct components that both filter based on different data and requirements :)

Basically, it comes down to what the point of the review sections is. Is it "give me exercises tailored to my current level of understanding so I can practice this subject matter" or "give me exercises that I should expect to see in quizzes, homeworks, exams, etc."? Because if it's the former - whether they came up in homeworks/exams/quizzes doesn't matter at all; it's entirely formative. 
If it's the latter, then I, as a student, don't *want* any additional filtering, on the contrary - if there are exam-relevant exercises, I want to see them whether my user model satisfies the prerequisites or not, because *especially* if it doesn't, that is highly relevant information for me.

One can of course just split the review section in two:
1. Here are former quiz/homework/exam questions on this particular topic, and
2. Here are *additional* exercises base on your user model

the former only requires BoF-metadata, the latter only requires the UMP, and then we just subtract 1. from 2.

**Comment by kohlhase** [comment]
I have been thinking along the same lines. Splitting along the learning usecases seems reasonable thing. We may have to find out what reasonable learning usecases actually are by talking to students.

**Comment by lambdaTotoro** [comment]
Late to the party, but I agree with Dennis' assessment that we (read: the students) probably want both separately and not at once because learning lecture material at my own pace in my own way and oh-my-various-gods-the-exam-is-tomorrow-cramming are both realistic use cases we want to support.

**Comment by abchugh** [comment]
Based on our discussion today, there is no clean way to this currently. Changes to the JSON file cannot be made: (a) the JSON producing code in MMT was a hack that is not worth delving into at this time. (b) MMT doesn't have a notion of quizzes/exams etc. 

So for now, Michael & Fredreick will work together (grep) and provide a list of problems where each problem will be  marked if they were on previous, quizzes, exams, or homework. BoF will store this data and use it to create appropriate  problem sets marked as "Quiz Problems", "Exam Problems" etc. on the section review interface
